File: JTSArjuna/JTSArjuna-2.0_Alpha_2_1/ArjunaCommon/classes/com/arjuna/ArjunaCommon/Common/ObjectName.java
/*
 * Copyright (C) 1996, 1997, 1998, 1999, 2000,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: ObjectName.java,v 1.1 2000/02/25 14:02:38 nmcl Exp $
 */

package com.arjuna.ArjunaCommon.Common;

import com.arjuna.ArjunaCommon.Interface.NameService;
import java.io.PrintStream;

import java.lang.NullPointerException;
import java.io.IOException;

public class ObjectName
{
    /*
     * All numbers are signed, but we leave this in for compatibility.
     */
    
public static final int SIGNED_NUMBER = 0;
public static final int UNSIGNED_NUMBER = 1;
public static final int STRING = 2;
public static final int OBJECTNAME = 3;
public static final int CLASSNAME = 4;
public static final int UID = 5;

public ObjectName (String objectName)
    {
	_mech = null;
	_name = null;
	_nameService = null;

	if (objectName == null)
	{
	    _mech = _invalidString;
	    _name = _invalidString;
	}
	else
	{
	    int mechLength = objectName.indexOf(nameDelim);

	    if (mechLength != -1)
	    {
		int nameLength = objectName.length() - mechLength - nameDelim.length();

		_mech = new String(objectName.substring(0, mechLength));

		if (nameLength > 0)
		    _name = objectName.substring((mechLength+nameDelim.length()), objectName.length());
		else
		    _name = new String("");
	    }
	}

	initNameService();
    }

public ObjectName (ObjectName objectName)
    {
	_mech = null;
	_name = null;
	_nameService = null;

	this.copy(objectName);
    }

public ObjectName (NameService nameService, String objectName)
    {
	_mech = null;
	_name = null;
	_nameService = null;
	
	if ((nameService != null) && (objectName != null))
	{
	    _nameService = nameService;

	    String fullMechName = _nameService.impleClassName().stringForm();
	    int appendedString = fullMechName.indexOf(nameServiceString);

	    _mech = new String(fullMechName.substring(0, appendedString-fullMechName.length()));
	    _name = new String(objectName);
	}
	else
	{
	    if (nameService == null)
		System.out.println("ERROR: ObjectName - no name service given.");
	
	    if (objectName == null)
		System.out.println("ERROR: ObjectName - no object name give.");

	    _mech = _invalidString;
	    _name = _invalidString;
	}
    }

public void finalize ()
    {
	_mech = null;
	_name = null;
	_nameService = null;
    }
    
public int attributeType (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.attributeType(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public String firstAttributeName () throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.firstAttributeName(_name);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public String nextAttributeName (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.nextAttributeName(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public long getLongAttribute (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.getLongAttribute(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public String getStringAttribute (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.getStringAttribute(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public ObjectName getObjectNameAttribute (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.getObjectNameAttribute(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public ClassName getClassNameAttribute (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.getClassNameAttribute(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public Uid getUidAttribute (String attrName) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.getUidAttribute(_name, attrName);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public void setLongAttribute (String attrName, long value) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    _name = _nameService.setLongAttribute(_name, attrName, value);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public void setStringAttribute (String attrName, String value) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    _name = _nameService.setStringAttribute(_name, attrName, value);
	else
	    throw(new NullPointerException("NameService implementation not set"));	
    }

public void setObjectNameAttribute (String attrName, ObjectName value) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    _name = _nameService.setObjectNameAttribute(_name, attrName, value);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public void setClassNameAttribute (String attrName, ClassName value) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    _name = _nameService.setClassNameAttribute(_name, attrName, value);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public void setUidAttribute (String attrName, Uid value) throws NullPointerException, IOException
    {
	if (_nameService != null)
	    _name = _nameService.setUidAttribute(_name, attrName, value);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public boolean removeAttribute (String attrName) throws IOException
    {
	if (_nameService != null)
	{
	    _name = _nameService.removeAttribute(_name, attrName);

	    if (_name != null)
		return true;
	}
	
	return false;
    }

public String uniqueAttributeName () throws NullPointerException, IOException
    {
	if (_nameService != null)
	    return _nameService.uniqueAttributeName(_name);
	else
	    throw(new NullPointerException("NameService implementation not set"));
    }

public static ObjectName uniqueObjectName (String mech) throws NullPointerException, IOException
    {
	NameService nameService = null;
	ObjectName objectName = ObjectName.invalid();

	if ((mech != null) && (mech.compareTo(_invalidString) != 0))
	{
	    String nameServiceMech = mech + nameServiceString;
	    ClassName nameServiceMechClassName = new ClassName(nameServiceMech);
	    nameService = NameService.create(nameServiceMechClassName);

	    if (nameService != null)
		objectName = nameService.uniqueObjectName();
	    
	    nameService = null;
	}

	return objectName;
    }

public synchronized void copy (ObjectName objectName)
    {
	if (objectName == this)
	    return;
	
	_nameService = null;
        _mech = null;
        _name = null;

	if (objectName._mech != null)
	{
	    _mech = new String(objectName._mech);

	    if (objectName._nameService != null)
		_nameService = new NameService(objectName._nameService);
	}

	if (objectName._name != null)
	    _name = new String(objectName._name);

	initNameService();
    }

public boolean equals (ObjectName objectName)
    {
	return (((_name == objectName._name) || 
		((_name != null) && (objectName._name != null) && (_name.compareTo(objectName._name) == 0))) &&
	    ((_mech == objectName._mech) ||
	     ((_mech != null) && (objectName._mech != null) && (_mech.compareTo(objectName._mech) == 0))));
    }

public boolean notEquals (ObjectName objectName)
    {
	return (((_name != objectName._name) && 
		((_name == null) || (objectName._name == null) || (_name.compareTo(objectName._name) != 0))) ||
	    ((_mech != objectName._mech) &&
	     ((_mech == null) || (objectName._mech == null) || (_mech.compareTo(objectName._mech) != 0))));
    }

public String getMechanismName ()
    {
	return _mech;
    }

public String stringForm ()
    {
	return new String(_mech + nameDelim + _name);
    }

public static ObjectName invalid ()
    {
	return _invalid;
    }

public void print (PrintStream strm)
    {
	strm.print("<ObjectName:"+_mech+nameDelim+_name+">");
    }

public String toString ()
    {
	return stringForm();
    }

public static void print (int t)
    {
	switch (t)
	{
	case ObjectName.SIGNED_NUMBER:
	    System.out.print("SIGNED_NUMBER");
	    break;
	case ObjectName.UNSIGNED_NUMBER:
	    System.out.print("UNSIGNED_NUMBER");
	    break;
	case ObjectName.STRING:
	    System.out.print("STRING");
	    break;
	case ObjectName.OBJECTNAME:
	    System.out.print("OBJECTNAME");
	    break;
	case ObjectName.CLASSNAME:
	    System.out.print("CLASSNAME");
	    break;
	case ObjectName.UID:
	    System.out.print("UID");
	    break;
	}
    }
    
private void initNameService ()
    {
	if (_nameService != null)
	    return;

	if ((_mech != null) && (_mech.compareTo(_invalidString) != 0))
	{
	    String nameServiceMech = new String(_mech + nameServiceString);
	    ClassName nameServiceMechClassName = new ClassName(nameServiceMech);
	    
	    _nameService = NameService.create(nameServiceMechClassName);
	    
	    nameServiceMech = null;
	}
	else
	    _nameService = null;
    }

private String _mech;
private String _name;
private NameService _nameService;
    
private static final String _invalidString = "invalid";
private static final String nameServiceString = "NameServiceImple";
private static final String nameDelim = ":";
private static final String nullString = "";
private static final ObjectName _invalid = new ObjectName("invalid:invalid");

}

File: JTSArjuna/JTSArjuna-1.2.2/Gandiva/lib/Common/ObjectName.cc
/*
 * Copyright (C) 1994, 1995, 1996,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: ObjectName.cc,v 1.10 1997/06/02 10:27:23 nmcl Exp $
 */

#include <os/stddef.h>
#include <os/string.h>
#include <os/iostream.h>

#ifndef COMMON_UID_H_
#   include <Common/Uid.h>
#endif

#ifndef COMMON_UTILITY_H_
#  include <Common/Utility.h>
#endif

#ifndef COMMON_OBJECTNAME_H_
#   include <Common/ObjectName.h>
#endif

#ifndef COMMON_CLASSNAME_H_
#   include <Common/ClassName.h>
#endif

#ifndef INTERFACE_BUFFER_H_
#   include <Interface/Buffer.h>
#endif

#ifndef INTERFACE_NAMESERVICE_H_
#   include <Interface/NameService.h>
#endif

ObjectName* ObjectName::_invalid = (ObjectName*) 0;

static const char* _invalidString = "invalid";
static const char* nameServiceString = "NameServiceImple";

const char* nameDelim = ":";
const char* nullString = "";


Boolean ObjectName::attributeType(const char *attrName,
				  AttributeType &attrType)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->attributeType(_name, attrName, attrType);

    return res;
}

Boolean ObjectName::firstAttributeName(char *&attrName)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->firstAttributeName(_name, attrName);

    return res;
}

Boolean ObjectName::nextAttributeName(const char *attrName,
				      char *&nextAttrName)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->nextAttributeName(_name, attrName, nextAttrName);

    return res;
}

Boolean ObjectName::getAttribute(const char *attrName, long &value) const
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
        cerr << "WARNING: ObjectName::getAttribute(\"" << attrName
	     << "\", long&) on " <<  *this << " failed" << endl;

    return res;
}

Boolean ObjectName::getAttribute(const char *attrName,
				 unsigned long &value) const
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
        cerr << "WARNING: ObjectName::getAttribute(\"" << attrName
	     << "\", unsigned long&) on " <<  *this << " failed" << endl;

    return res;
}

Boolean ObjectName::getAttribute(const char *attrName, char *&value) const
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
        cerr << "WARNING: ObjectName::getAttribute(\"" << attrName
	     << "\", char*&) on " <<  *this << " failed" << endl;

    return res;
}

Boolean ObjectName::getAttribute(const char *attrName, ObjectName &value) const
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
        cerr << "WARNING: ObjectName::getAttribute(\"" << attrName
	     << "\", ObjectName&) on " <<  *this << " failed" << endl;

    return res;
}

Boolean ObjectName::getAttribute(const char *attrName, ClassName &value) const
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
        cerr << "WARNING: ObjectName::getAttribute(\"" << attrName
	     << "\", ClassName&) on " <<  *this << " failed" << endl;

    return res;
}

Boolean ObjectName::getAttribute(const char *attrName, Uid &value) const
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
        cerr << "WARNING: ObjectName::getAttribute(\"" << attrName
	     << "\", Uid&) on " <<  *this << " failed" << endl;

    return res;
}

ObjectName ObjectName::getObjectNameAttribute(const char *attrName) const
{
    Boolean res = FALSE;
    ObjectName value(ObjectName::invalid());

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
    {
        cerr << "WARNING: ObjectName::getObjectNameAttribute(\""
	     << attrName << "\") on " <<  *this << " failed" << endl;

        value = ObjectName::invalid();
    }

    return value;
}

ClassName ObjectName::getClassNameAttribute(const char *attrName) const
{
    Boolean res = FALSE;
    ClassName value(ClassName::invalid());

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
    {
        cerr << "WARNING: ObjectName::getClassNameAttribute(\""
	     << attrName << "\") on " <<  *this << " failed" << endl;

        value = ClassName::invalid();
    }

    return value;
}

Uid ObjectName::getUidAttribute(const char *attrName) const
{
    Boolean res = FALSE;
    Uid value(Uid::invalid());

    if (_nameService != NULL)
        res = _nameService->getAttribute(_name, attrName, value);

    if (! res)
    {
        cerr << "WARNING: ObjectName::getUidAttribute(\"" << attrName
	     << "\") on " <<  *this << " failed" << endl;

        value = Uid::invalid();
    }

    return value;
}

Boolean ObjectName::setAttribute(const char *attrName, long value)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->setAttribute(_name, attrName, value);

    return res;
}

Boolean ObjectName::setAttribute(const char *attrName, unsigned long value)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->setAttribute(_name, attrName, value);

    return res;
}

Boolean ObjectName::setAttribute(const char *attrName, const char *value)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->setAttribute(_name, attrName, value);

    return res;
}

Boolean ObjectName::setAttribute(const char *attrName, const ObjectName &value)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->setAttribute(_name, attrName, value);

    return res;
}

Boolean ObjectName::setAttribute(const char *attrName, const ClassName &value)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->setAttribute(_name, attrName, value);

    return res;
}

Boolean ObjectName::setAttribute(const char *attrName, const Uid &value)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->setAttribute(_name, attrName, value);

    return res;
}

Boolean ObjectName::removeAttribute(const char *attrName)
{
    Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->removeAttribute(_name, attrName);

    return res;
}

Boolean ObjectName::uniqueAttributeName(char *&attrName)
{
   Boolean res = FALSE;

    if (_nameService != NULL)
        res = _nameService->uniqueAttributeName(_name, attrName);

    return res;
}

Boolean ObjectName::uniqueObjectName(const char *mech, ObjectName &objectName)
{
    Boolean res = FALSE;

    NameService *nameService = NULL;

    objectName = ObjectName::invalid();

    if ((mech != NULL) && (strcmp(mech, _invalidString) != 0))
    {
        char *nameServiceMech = new char[strlen(mech) + strlen(nameServiceString) + 1];

        strcpy(nameServiceMech, mech);
        strcat(nameServiceMech, nameServiceString);

        ClassName nameServiceMechClassName(nameServiceMech);
        nameService = NameService::create(nameServiceMechClassName);

        Resource::ref(nameService);

        if (nameService != NULL)
            res = nameService->uniqueObjectName(objectName);

        if (nameServiceMech != NULL)
            ::delete [] nameServiceMech;

        Resource::unref(nameService);
    }

    return res;
}

ObjectName &ObjectName::operator= (const ObjectName &objectName)
{
    if (this == &objectName)
	return *this;
    
    Resource::unref(_nameService);
    _nameService = NULL;

    if (_mech != NULL)
    {
        ::delete [] _mech;
        _mech = NULL;
    }

    if (_name != NULL)
    {
        ::delete [] _name;
        _name = NULL;
    }

    if (objectName._mech != NULL)
    {
        _mech = new char[strlen(objectName._mech) + 1];
        strcpy(_mech, objectName._mech);
	if (objectName._nameService)
	{
	    _nameService = new NameService(*objectName._nameService);
	    Resource::ref(_nameService);
	}
    }

    if (objectName._name != NULL)
    {
        _name = new char[strlen(objectName._name) + 1];
        strcpy(_name, objectName._name);
    }

    initNameService();

    return *this;
}

Boolean ObjectName::operator== (const ObjectName &objectName) const
{
    return ((_name == objectName._name) || 
            ((_name != NULL) && (objectName._name != NULL) && (strcmp(_name, objectName._name) == 0))) &&
           ((_mech == objectName._mech) ||
            ((_mech != NULL) && (objectName._mech != NULL) && (strcmp(_mech, objectName._mech) == 0)));
}

Boolean ObjectName::operator!= (const ObjectName &objectName) const
{
    return ((_name != objectName._name) && 
            ((_name == NULL) || (objectName._name == NULL) || (strcmp(_name, objectName._name) != 0))) ||
           ((_mech != objectName._mech) &&
            ((_mech == NULL) || (objectName._mech == NULL) || (strcmp(_mech, objectName._mech) != 0)));
}

ObjectName::ObjectName(const char *objectName)
		      : _mech(0),
			_name(0),
			_nameService(0)
{
    if (objectName == NULL)
    {
       _mech = ::new char [::strlen(_invalidString) + 1];
       ::strcpy(_mech, _invalidString);
       _name = ::new char [::strlen(_invalidString) + 1];
       ::strcpy(_name, _invalidString);
    }
    else
    {
        char *colon = ::strstr(objectName, nameDelim);
    
        if (colon != NULL)
        {
            int mechLength = colon - objectName;
            int nameLength = ::strlen(objectName) - mechLength - ::strlen(nameDelim);

	    _mech = ::new char[mechLength + 1];

	    ::strncpy(_mech, objectName, mechLength);
	    _mech[mechLength] = '\0';
	    
	    if (nameLength > 0)
	    {
		_name = ::new char[nameLength + 1];

		::strncpy(_name, colon + ::strlen(nameDelim), nameLength);
		_name[nameLength] = '\0';
	    }
	    else
	    {
		_name = ::new char[::strlen(nullString)+1];
		::strcpy(_name, nullString);
	    }
	}
    }

    initNameService();
}

ObjectName::ObjectName(const ObjectName &objectName)
		      : _mech(0),
			_name(0),
			_nameService(0)
{
    if (objectName._mech != NULL)
    {
        _mech = ::new char[strlen(objectName._mech) + 1];
        ::strcpy(_mech, objectName._mech);
	if (objectName._nameService)
	{
	    _nameService = new NameService(*objectName._nameService);
	    Resource::ref(_nameService);
	}
    }
    else
    {
       _mech = ::new char [::strlen(_invalidString) + 1];
       ::strcpy(_mech, _invalidString);
    }

    if (objectName._name != NULL)
    {
        _name = ::new char[::strlen(objectName._name) + 1];
        ::strcpy(_name, objectName._name);
    }
    else
    {
       _name = ::new char [::strlen(_invalidString) + 1];
       ::strcpy(_name, _invalidString);
    }

    initNameService();
}

ObjectName::ObjectName (NameService *nameService, const char *objectName)
		       : _mech(0),
			 _name(0),
			 _nameService(0)
{
    if ((nameService) && (objectName))
    {
	_nameService = nameService;
	Resource::ref(_nameService);

	char* fullMechName = _nameService->impleClassName().stringForm();
	char* appendedString = ::strstr(fullMechName, nameServiceString);
	
	_mech = ::new char[appendedString - fullMechName +1];
	::strncpy(_mech, fullMechName, appendedString-fullMechName);
	_mech[appendedString-fullMechName] = '\0';
	::delete [] fullMechName;

	_name = ::new char[::strlen(objectName)+1];
	::strcpy(_name, objectName);
    }
    else
    {
	if (!nameService)
	    cerr << "ERROR: ObjectName::ObjectName - no name service given."
		 << endl;
	
	if (!objectName)
	    cerr << "ERROR: ObjectName::ObjectName - no object name give."
		 << endl;
	
       _mech = ::new char [::strlen(_invalidString) + 1];
       ::strcpy(_mech, _invalidString);
       _name = ::new char [::strlen(_invalidString) + 1];
	::strcpy(_name, _invalidString);
    }
}

ObjectName::~ObjectName()
{
    if (_mech != NULL)
        ::delete [] _mech;

    if (_name != NULL)
        ::delete [] _name;

    Resource::unref(_nameService);
}

char* ObjectName::getMechanismName () const
{
    if (!_mech)
        return (char*) 0;
    
    char* ptr = ::new char[::strlen(_mech)+1];
    ::strcpy(ptr, _mech);

    return ptr;
}

char *ObjectName::stringForm () const
{
    char* ptr1 = ((_mech) ? ::appendStrings(_mech, nameDelim) : 0);
    char* ptr2 = ::appendStrings(ptr1, _name);
    
    if (ptr1)
	::delete [] ptr1;

    return ptr2;
}

const ObjectName &ObjectName::invalid ()
{
    if (!_invalid)
	_invalid = new ObjectName("invalid:invalid");
    
    return *_invalid;
}

Boolean ObjectName::pack(Buffer &buff) const
{
    return buff.pack(_mech) && buff.pack(_name);
}

Boolean ObjectName::unpack(Buffer &buff)
{
    Boolean res = TRUE;

    Resource::unref(_nameService);
    _nameService = NULL;

    res = buff.unpack(_mech) && buff.unpack(_name);

    if (res)
        initNameService();

    return res;
}

void ObjectName::print(ostream &ostr) const
{
    ostr << "<ObjectName:" << _mech << nameDelim << _name << '>';
}

void ObjectName::initNameService()
{
    if (_nameService)
	return;

    if ((_mech != NULL) && (::strcmp(_mech, _invalidString) != 0))
    {
	char* nameServiceMech = ::appendStrings(_mech, nameServiceString);

        ClassName nameServiceMechClassName(nameServiceMech);

        _nameService = NameService::create(nameServiceMechClassName);

        Resource::ref(_nameService);

        if (nameServiceMech != NULL)
            ::delete [] nameServiceMech;
    }
    else
        _nameService = NULL;
}

ostream &operator<< (ostream &ostr, const ObjectName &objectName)
{
    objectName.print(ostr);

    return ostr;
}

ostream& operator<< (ostream& strm, const ObjectName::AttributeType t)
{
    switch (t)
    {
    case ObjectName::SIGNED_NUMBER:
	strm << "SIGNED_NUMBER";
	break;
    case ObjectName::UNSIGNED_NUMBER:
	strm << "UNSIGNED_NUMBER";
	break;
    case ObjectName::STRING:
	strm << "STRING";
	break;
    case ObjectName::OBJECTNAME:
	strm << "OBJECTNAME";
	break;
    case ObjectName::CLASSNAME:
	strm << "CLASSNAME";
	break;
    case ObjectName::UID:
	strm << "UID";
	break;
    }

    return strm;
}
